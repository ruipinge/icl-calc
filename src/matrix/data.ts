import { DATA_POINTS, DataPoint } from '../db';

const ATA_DEVIATION: number = 0.1; // millimetres
const CLR_DEVIATION: number = 100 * 0.001; // millimetres

export const NUM_DATA_POINTS = DATA_POINTS.length;

export interface VaultRange {
  min?: number;
  max?: number;
}

type LensSizeId = 'small' | 'medium' | 'large';
interface LensSize {
  id: LensSizeId;
  label: string;
  value: number;
}
export const LENS_SIZES: LensSize[] = [
  {
    label: '12.6 mm',
    id: 'small',
    value: 12.6
  },
  {
    label: '13.2 mm',
    id: 'medium',
    value: 13.2
  },
  {
    label: '13.7 mm',
    id: 'large',
    value: 13.7
  }
];

type MyopiaId = 'low' | 'moderate' | 'high';
interface MyopiaLevel {
  id: MyopiaId;
  label: string;
  title: string;
  min: number;
  max: number;
}
export const MYOPIA_LEVELS: MyopiaLevel[] = [
  {
    id: 'low',
    label: 'Low',
    title: 'Spherical Equivalent (SE) less than -6 dpt',
    min: -6,
    max: Infinity
  },
  {
    id: 'moderate',
    label: 'Moderate',
    title: 'Spherical Equivalent (SE) between -6 dpt and -12 dpt',
    min: -12,
    max: -6
  },
  {
    id: 'high',
    label: 'High',
    title: 'Spherical Equivalent (SE) greater than -12 dpt',
    min: -Infinity,
    max: -12
  }
];

export interface MatrixFilter {
  readonly ata: number;
  readonly clr: number;
  readonly dataPoints?: DataPoint[];
}

interface CacheEntry {
  lens: LensSize;
  myopia: MyopiaLevel;
  dataPoints?: DataPoint[];
}

export const getDataPoints = ({
  lens,
  myopia,
  dataPoints = DATA_POINTS
}: CacheEntry): DataPoint[] =>
  dataPoints.filter(
    (point) =>
      lens.value === point.iclSize &&
      myopia.min <= point.iclSe &&
      point.iclSe < myopia.max
  );

const filter = ({
  ata,
  clr,
  point
}: {
  ata: number;
  clr: number;
  point: DataPoint;
}) =>
  ata >= point.ata - ATA_DEVIATION &&
  ata <= point.ata + ATA_DEVIATION &&
  clr >= point.clr - CLR_DEVIATION && // clr in db is in millimetres
  clr <= point.clr + CLR_DEVIATION; // clr in db is in millimetres

export const filterFlatRows = ({
  ata,
  clr,
  dataPoints = DATA_POINTS
}: MatrixFilter): DataPoint[][] =>
  LENS_SIZES.reduce(
    (memoLens, lens) => [
      ...memoLens,
      ...MYOPIA_LEVELS.reduce(
        (memoMyopia, level) => [
          ...memoMyopia,
          ...[
            getDataPoints({
              lens: lens,
              myopia: level,
              dataPoints: dataPoints
            }).filter((point: DataPoint) => filter({ ata, clr, point }))
          ]
        ],
        [] as DataPoint[][]
      )
    ],
    [] as DataPoint[][]
  );
