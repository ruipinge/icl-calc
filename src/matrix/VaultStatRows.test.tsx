import {
  VaultStatRows,
  getVaultAverages,
  getVaultMaxs,
  getVaultMins
} from './VaultStatRows';

import TestRenderer from 'react-test-renderer';
import { FILTER as filter } from './data.test';

it('renders without crashing', () => {
  const tree = TestRenderer.create(<VaultStatRows ata={0} clr={0} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('calculates vault size averages', () => {
  const avgs = getVaultAverages(filter);
  expect(avgs).toEqual([1.2, 1.5, 1.8, 2.1, 2.4, 2.7, 3.0, 3.3, 3.6]);
});

it('calculates vault size minimums', () => {
  const avgs = getVaultMins(filter);
  expect(avgs).toEqual([1.1, 1.4, 1.7, 2.0, 2.3, 2.6, 2.9, 3.2, 3.5]);
});

it('calculates vault size maximums', () => {
  const avgs = getVaultMaxs(filter);
  expect(avgs).toEqual([1.3, 1.6, 1.9, 2.2, 2.5, 2.8, 3.1, 3.4, 3.7]);
});
