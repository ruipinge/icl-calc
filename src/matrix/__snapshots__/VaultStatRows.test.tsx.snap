// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
Array [
  <tr>
    <th
      scope="row"
      title="Average Vault size in micrometres"
    >
      Average Vault (μm)
    </th>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
  </tr>,
  <tr>
    <th
      scope="row"
      title="Minimum Vault size in micrometres"
    >
      Minimum Vault (μm)
    </th>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
  </tr>,
  <tr>
    <th
      scope="row"
      title="Maximum Vault size in micrometres"
    >
      Maximum Vault (μm)
    </th>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
  </tr>,
]
`;
